build-info-extractor-gradle breaks with Gradle 6.x when Gradle Module Metadata publication is disabled

Hi folks,

Gradle Module Metadata is a format used to serialize the Gradle component model that was introduced as default in Gradle 6.x. However, still possible to disable this for multiple reasons: https://docs.gradle.org/current/userguide/publishing_gradle_module_metadata.html#sub:disabling-gmm-publication

For example

```
tasks.withType(GenerateModuleMetadata) {
        enabled = false
    }
```

When this is disabled, 4.11.0 version of the plugin fails to publish with exceptions such as:

```Caused by: java.lang.RuntimeException: java.util.concurrent.ExecutionException: org.gradle.api.GradleException: File '/agent/workspace/NEBULA-nebula/subprojects/core/build/nebulatest/netflix.nebula.publishing.NebulaReleasePublishLauncherSpec/publish-a-jar/build/publications/nebulaIvy/module.json' does not exist, and need to be published from publication nebulaIvy```

While debugging found that ivyPublicationInternal.getPublishableArtifacts() has the module.json file on it

```
0 = {ArchiveTaskBasedIvyArtifact_Decorated@16692} "ArchiveTaskBasedIvyArtifact_Decorated testnebulareleasepublish:sources:jar:sources"
1 = {PublishArtifactBasedIvyArtifact_Decorated@17562} "PublishArtifactBasedIvyArtifact_Decorated testnebulareleasepublish:jar:jar:null"
2 = {SingleOutputTaskIvyArtifact@17563} "SingleOutputTaskIvyArtifact ivy:ivy:xml:null"
3 = {SingleOutputTaskIvyArtifact@17564} "SingleOutputTaskIvyArtifact testnebulareleasepublish:json:module:null"
```

This was introduced with https://github.com/jfrog/build-info/pull/265/files. 

Ideally, the plugin shouldn't try to publish files that are not being generated

Opening this ticket for visibility and already gave a heads up to @ljacomet

--- a/build-info-extractor-gradle/src/main/groovy/org/jfrog/gradle/plugin/artifactory/task/helper/TaskHelperPublications.java
+++ b/build-info-extractor-gradle/src/main/groovy/org/jfrog/gradle/plugin/artifactory/task/helper/TaskHelperPublications.java
@@ -27,7 +27,6 @@ import org.gradle.api.logging.Logging;
 import org.gradle.api.publish.Publication;
 import org.gradle.api.publish.PublicationArtifact;
 import org.gradle.api.publish.PublishingExtension;
-import org.gradle.api.publish.internal.PublicationArtifactSet;
 import org.gradle.api.publish.internal.PublicationInternal;
 import org.gradle.api.publish.ivy.IvyArtifact;
 import org.gradle.api.publish.ivy.IvyPublication;
@@ -214,7 +213,7 @@ public class TaskHelperPublications extends TaskHelper {
             }
 
             // Second adding all artifacts, skipping the ivy file
-            PublicationArtifactSet<IvyArtifact> artifacts = ivyPublicationInternal.getPublishableArtifacts();
+            Set<IvyArtifact> artifacts = ivyNormalizedPublication.getAllArtifacts();
             for (IvyArtifact artifact : artifacts) {
                 File file = artifact.getFile();
                 // Skip the ivy file

--- a/build-info-extractor-gradle/src/main/groovy/org/jfrog/gradle/plugin/artifactory/task/helper/TaskHelperPublications.java
+++ b/build-info-extractor-gradle/src/main/groovy/org/jfrog/gradle/plugin/artifactory/task/helper/TaskHelperPublications.java
@@ -239,16 +239,17 @@ public class TaskHelperPublications extends TaskHelper {
             MavenNormalizedPublication mavenNormalizedPublication = mavenPublicationInternal.asNormalisedPublication();
 
             // First adding the Maven descriptor (if the build is configured to add it):
+            File pomFile = mavenNormalizedPublication.getPomArtifact().getFile();
             if (isPublishMaven()) {
-                File file = mavenNormalizedPublication.getPomArtifact().getFile();
-                DeployDetails.Builder builder = createBuilder(file, publicationName);
+                DeployDetails.Builder builder = createBuilder(pomFile, publicationName);
                 if (builder != null) {
                     PublishArtifactInfo artifactInfo = new PublishArtifactInfo(
-                            mavenPublication.getArtifactId(), "pom", "pom", null, file);
+                            mavenPublication.getArtifactId(), "pom", "pom", null, pomFile);
                     addMavenArtifactToDeployDetails(deployDetails, publicationName, builder, artifactInfo, mavenPublication);
                 }
             }
 
+            boolean legacy = false;
             Set<MavenArtifact> artifacts;
             try {
                 // Gradle 5.0 and above:
@@ -260,10 +261,15 @@ public class TaskHelperPublications extends TaskHelper {
             } catch (NoSuchMethodError error) {
                 // Compatibility with older versions of Gradle:
                 artifacts = mavenNormalizedPublication.getAllArtifacts();
+                legacy = true;
             }
 
             // Third adding all additional artifacts - includes Gradle Module Metadata when produced
             for (MavenArtifact artifact : artifacts) {
+                if (legacy && artifact.getFile().equals(pomFile)) {
+                    // Need to skip the POM file for Gradle < 5.0
+                    continue;
+                }
                 createPublishArtifactInfoAndAddToDeployDetails(artifact, deployDetails, mavenPublication, publicationName);
             }
         }

